### PR TITLE
Get proper renderer width and height in FigureImage

### PR DESCRIPTION
**Pull request by jklymak**
Possible fix to https://github.com/matplotlib/matplotlib/issues/9203.  Works with their example. 

Issue was that `renderer.width` would return 1.0, but we really needed `width, height = renderer.get_canvas_width_height()`.

Their test was:

```python

import matplotlib.pyplot as plt
import numpy as np
data = np.arange(100*100).reshape(100, 100).astype(float)
plt.imsave("test.pdf", data, origin="lower", format="pdf") #same behavior also with eps instead of pdf

```

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant

### CODE REVIEW

**Comment by dstansby** [comment]
This definitely needs a test adding! Could you add your example in the bug report to `lib/matplotlib/tests/test_image.py`?

**Comment by jklymak** [comment]
Changig to WIP - see below, but `width, height = renderer.get_canvas_width_height()` doesn't seem to know about `dpi` in `fig = plt.figure(figsize=(2,2), dpi=100)`...

@dstansby 

For the tests, there is already the test below, but as you can see it is only set to check `png`.  

```python

@image_comparison(baseline_images=['figimage-0', 'figimage-1'], extensions=['png'])
def test_figimage():
    'test the figimage method'

    for suppressComposite in False, True:
        fig = plt.figure(figsize=(2,2), dpi=100)
        fig.suppressComposite = suppressComposite
        x,y = np.ix_(np.arange(100.0)/100.0, np.arange(100.0)/100.0)
        z = np.sin(x**2 + y**2 - x*y)
        c = np.sin(20*x**2 + 50*y**2)
        img = z + c/5

        fig.figimage(img, xo=0, yo=0, origin='lower')
        fig.figimage(img[::-1,:], xo=0, yo=100, origin='lower')
        fig.figimage(img[:,::-1], xo=100, yo=0, origin='lower')
        fig.figimage(img[::-1,::-1], xo=100, yo=100, origin='lower')
```

Unfortunately this turns up a continued error, and I'm not quite sure how to get around it...

If I modify the above to run from the command line:

```python

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import numpy as np
for suppressComposite in False, True:
    fig = plt.figure(figsize=(2,2), dpi=100)
    fig.suppressComposite = suppressComposite
    x,y = np.ix_(np.arange(100.0)/100.0, np.arange(100.0)/100.0)
    z = np.sin(x**2 + y**2 - x*y)
    c = np.sin(20*x**2 + 50*y**2)
    img = z + c/5

    fig.figimage(img, xo=0, yo=0, origin='lower')
    fig.figimage(img[::-1,:], xo=0, yo=100, origin='lower')
    fig.figimage(img[:,::-1], xo=100, yo=0, origin='lower')
    fig.figimage(img[::-1,::-1], xo=100, yo=100, origin='lower')

    if suppressComposite:
        fig.savefig('Boo.pdf')
        fig.savefig('Boo.png')
    else:
        fig.savefig('Boo2.pdf')
        fig.savefig('Boo2.png')
```

For `pdf` rendering `get_canvas_width_height()` is returning `144, 144` instead of `200, 200` for the above so it is ignoring the figure specified `dpi` and using a default value of `72`.  If I save as `png`, it assumes that `dpi=300` and returns `600, 600`.  

So, somehow the backends are getting the `figsize`, but they are not getting the `dpi` from the   `fig = plt.figure(figsize=(2,2), dpi=100)` command, at least as reported by ` width, height = renderer.get_canvas_width_height()`

**Comment by jklymak** [comment]
OK, so the `matplotlib.backends.backend_mixed.MixedModeRenderer`, which gets used when making the pdf and the `matplotlib.backends.backend_agg.RendererAgg` return very different values of `width` and `height`, and that is the problem.

For `renderer.width`, `RenderAgg` returns `fig.width * fig.dpi` and `MixedModeRenderer` simply returns `fig.width`.  So, this PR is almost certainly wrong to use the canvas method, but the two renderers should probably be made compatible...

**Comment by jklymak** [comment]
@dstansby the test is the existing `test_image.py:test_figimage` with the `svg` and `pdf` filetypes added...

**Comment by jklymak** [comment]
Todo: remove or fix the svg test...

File: lib/matplotlib/tests/test_image.py [hunk 1]
```diff
@@ -72,11 +72,12 @@ def test_interp_nearest_vs_none():
...
 
-@image_comparison(baseline_images=['figimage-0', 'figimage-1'], extensions=['png'])
-def test_figimage():
+@image_comparison(baseline_images=['figimage-0'],
+                  extensions=['png','pdf'])
+def test_figimage0():
     'test the figimage method'
 
-    for suppressComposite in False, True:
+    for suppressComposite in [False]:
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_image.py]: I am a bit concerned that we are losing testing the `supressComposite=True` code path.

Comment (tacaswell) [reply lib/matplotlib/tests/test_image.py]: Oh, never mind, I should have read the next part of diff :sheep:

Comment (jklymak) [reply lib/matplotlib/tests/test_image.py]: Yeah, maybe I should clean this a bit. The for-loop worked when the test only did `png`, but never got to the second pdf for some reason.  This (ahem) "fixed" the problem.

**Comment by tacaswell** [comment]
Thanks! backported to v2.1.x as 052b52bf2

**Pull request merged**